python/ray/data/dataset.py

```diff
@@ -5918,6 +5928,117 @@ def explain(self):
+            ... ])
+            >>> summary = ds.summary()
+            >>> print(summary)  # doctest: +SKIP
+                agg         numerical               categorical       vector
+                column             age        salary        name  city scores  tags
+                count         4.000000      4.000000         4.0   4.0    4.0   4.0
+                mean         18.333333  51666.666667         NaN   NaN    NaN   NaN
+                min           0.000000  45000.000000         NaN   NaN    NaN   NaN
+                max          30.000000  60000.000000         NaN   NaN    NaN   NaN
+                std          13.123346   6236.095645         NaN   NaN    NaN   NaN
+                missing_pct  25.000000     25.000000        25.0  25.0   25.0  25.0
+                zero_pct     33.333333      0.000000         NaN   NaN    NaN   NaN
+
+        Args:
+            columns: Optional list of column names to include in the summary.
+                If None, all columns will be included.
+
+        Returns:
+            A DatasetSummary object containing separate DataFrames for each column
+            type (numerical, categorical, vector).
+        """
+        import pandas as pd
+
+        from ray.data.stats import (
+            STAT_ORDER,
+            ColumnType,
+            DatasetSummary,
+            feature_aggregators_for_dataset,
+            get_stat_names_for_column_type,
+        )
+
+        # Get aggregators and compute results
+        feature_aggs = feature_aggregators_for_dataset(self, columns=columns)
+        results = self.aggregate(*feature_aggs.aggregators)
+        if not results:
+            return DatasetSummary(
+                numerical=pd.DataFrame(),
+                categorical=pd.DataFrame(),
+                vector=pd.DataFrame(),
+            )
+
+        # Helper to extract stats for a column from aggregation results
+        def _extract_column_stats(column_name, stat_names):
+            """Extract statistics for a single column."""
+            column_stats = {}
+            for stat in stat_names:
+                # Aggregator names follow pattern: "stat_name(column_name)"
+                key = _make_aggregator_name(stat, column_name)
+                if key in results:
+                    column_stats[stat] = results[key]
+            return column_stats
+
+        # Build separate DataFrames for each column type
+        column_data_map = {
+            ColumnType.NUMERICAL: {},
+            ColumnType.CATEGORICAL: {},
+            ColumnType.VECTOR: {},
+        }
+
+        # Process all column types using enum iteration
+        for column_type in ColumnType:
+            stat_names = get_stat_names_for_column_type(column_type)
+            columns_for_type = feature_aggs.columns_by_type[column_type]
+
+            for col in columns_for_type:
+                column_data_map[column_type][col] = _extract_column_stats(
+                    col, stat_names
+                )
```

Comment:
We don't need all of that complexity, right?

1. Just group all of your columns by type
2. Keep col_type -> col -> list(aggs) map
3. Resolve its values here (by using agg.name directly)